cmake/SetupCompilerOptions.cmake

```diff
@@ -97,49 +97,130 @@ endif()
-    # Support Extra Flags for the C compiler.
-    ##########################################
-    if(EXTRA_C_FLAGS)
-        set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${EXTRA_C_FLAGS}")
-    endif()
 
-    # Extra Flags for the debug builds with the C compiler.
-    if(EXTRA_C_FLAGS_DEBUG AND
-      ( (CMAKE_BUILD_TYPE MATCHES Debug)
-        OR (CMAKE_BUILD_TYPE MATCHES RelWithDebInfo) )
-      )
-        add_compile_options("${EXTRA_C_FLAGS_DEBUG}")
-    endif()
+    # Extra Flags for debug builds 
+    if(CMAKE_BUILD_TYPE MATCHES Debug)
+        # debug flags for the C compiler
+        if(BLT_C_FLAGS_DEBUG)
+            set(CMAKE_C_FLAGS_DEBUG
+                "${CMAKE_C_FLAGS_DEBUG} ${BLT_C_FLAGS_DEBUG}")
+        endif()
+
+        # debug flags for the C++ compiler
+        if(BLT_CXX_FLAGS_DEBUG)
+            set(CMAKE_CXX_FLAGS_DEBUG
+                "${CMAKE_CXX_FLAGS_DEBUG} ${BLT_CXX_FLAGS_DEBUG}")
+        endif()
+
+        # debug flags for the Fortran compiler
+        if(ENABLE_FORTRAN AND BLT_FORTRAN_FLAGS_DEBUG)
+            set(CMAKE_Fortran_FLAGS_DEBUG
+                "${CMAKE_Fortran_FLAGS_DEBUG} ${BLT_FORTRAN_FLAGS_DEBUG}")
+        endif()
 
-    # Extra Flags for the release builds with the C compiler.
-    if(EXTRA_C_FLAGS_RELEASE AND CMAKE_BUILD_TYPE MATCHES RELEASE)
-        add_compile_options("${EXTRA_C_FLAGS_RELEASE}")
     endif()
 
-    #############################################
-    # Support Extra Flags for the C++ compiler.
-    #############################################
-    if(EXTRA_CXX_FLAGS)
-        add_compile_options("${EXTRA_CXX_FLAGS}")
+    # Extra Flags for release builds
+    if(CMAKE_BUILD_TYPE MATCHES RELEASE)
+
+        # release flags for the C compiler
+        if(BLT_C_FLAGS_RELEASE)
+            set(CMAKE_C_FLAGS_RELEASE
+                "${CMAKE_C_FLAGS_RELEASE} ${BLT_C_FLAGS_RELEASE}")
+        endif()
+
+        # release flags for the C++ compiler
+        if(BLT_CXX_FLAGS_RELEASE)
+            set(CMAKE_CXX_FLAGS_RELEASE
+                "${CMAKE_CXX_FLAGS_RELEASE} ${BLT_CXX_FLAGS_RELEASE}")
+        endif()
+
+        # release flags for the Fortran compiler.
+        if(ENABLE_FORTRAN AND BLT_FORTRAN_FLAGS_RELEASE)
+            set(CMAKE_Fortran_FLAGS_RELEASE
+                "${CMAKE_Fortran_FLAGS_RELEASE} ${BLT_FORTRAN_FLAGS_RELEASE}")
+        endif()
+        
     endif()
-
-    # Extra Flags for the debug builds with the C++ compiler.
-    if(EXTRA_CXX_FLAGS_DEBUG AND
-      ( (CMAKE_BUILD_TYPE MATCHES Debug)
-        OR (CMAKE_BUILD_TYPE MATCHES RelWithDebInfo) )
-      )
-        add_compile_options("${EXTRA_CXX_FLAGS_DEBUG}")
-    endif()
-
-    # Extra Flags for the release builds with the C++ compiler.
-    if(EXTRA_CXX_FLAGS_RELEASE AND CMAKE_BUILD_TYPE MATCHES RELEASE)
-        add_compile_options("${EXTRA_CXX_FLAGS_RELEASE}")
+    
+    # Extra Flags for release w/ debug info builds
+    if(CMAKE_BUILD_TYPE MATCHES RelWithDebInfo)
```

Comment:
I see (as a user) `RelWithDebInfo` as meaning the `Release` flags, plus whatever is needed to generate information for debuggers (e.g. `-g` for a host compiler, `-G` for CUDA).

Combining the `Release` and `Debug` flags might have consequences a user wouldn't expect.

I would propose that we set the flags to `Release` + `-g` by default, and have a new variable so that users can add any extra flags they want.